File: structural/flyweight/typescript/factory.ts
import { Circle } from "./concrete-flyweight";

export class ShapeFactory {
    private static circleMap: Map<string, Circle> = new Map();

    public static getCircle(color: string): Circle {
        let circle: Circle | undefined = ShapeFactory.circleMap.get(color);

        if (!circle) {
            circle = new Circle(color);
            ShapeFactory.circleMap.set(color, circle);
            console.log(`Creating circle of color : ${color}`);
        }
        return circle;
    }
}

File: structural/flyweight/python/factory.py
from typing import Dict

from concrete_flyweight import Circle


class ShapeFactory:
    _circle_map: Dict[str, Circle] = {}

    @staticmethod
    def get_circle(color: str) -> Circle:
        circle = ShapeFactory._circle_map.get(color)

        if not circle:
            circle = Circle(color)
            ShapeFactory._circle_map[color] = circle
            print(f"Creating circle of color : {color}")
        return circle
